component/stream/read_write.go

```diff
@@ -0,0 +1,458 @@
+package stream
+
+import (
+	"blobfuse2/common"
+	"blobfuse2/common/log"
+	"blobfuse2/internal"
+	"blobfuse2/internal/handlemap"
+	"encoding/base64"
+	"errors"
+	"io"
+	"strings"
+	"sync/atomic"
+
+	"github.com/pbnjay/memory"
+)
+
+type ReadWriteCache struct {
+	*Stream
+	StreamConnection
+}
+
+func (rw *ReadWriteCache) Configure(conf StreamOptions) error {
+	if conf.BufferSizePerFile <= 0 || conf.BlockSize <= 0 || conf.HandleLimit <= 0 {
+		rw.StreamOnly = true
+	}
+	rw.BlockSize = int64(conf.BlockSize) * mb
+	rw.BufferSizePerHandle = conf.BufferSizePerFile * mb
+	rw.HandleLimit = int32(conf.HandleLimit)
+	rw.CachedHandles = 0
+	return nil
+}
+
+func (rw *ReadWriteCache) CreateFile(options internal.CreateFileOptions) (*handlemap.Handle, error) {
+	log.Trace("Stream::CreateFile : name=%s, mode=%s", options.Name, options.Mode)
+	handle, err := rw.NextComponent().CreateFile(options)
+	if err != nil {
+		log.Err("Stream::CreateFile : error failed to create file %s: [%s]", options.Name, err.Error())
+	}
+	if !rw.StreamOnly {
+		err = rw.createHandleCache(handle)
+		if err != nil {
+			log.Err("Stream::CreateFile : error creating cache object %s [%s]", options.Name, err.Error())
+			return handle, err
+		}
+	}
+	return handle, nil
+}
+
+func (rw *ReadWriteCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
+	log.Trace("Stream::OpenFile : name=%s, flags=%d, mode=%s", options.Name, options.Flags, options.Mode)
+	handle, err := rw.NextComponent().OpenFile(options)
+	if err != nil {
+		log.Err("Stream::OpenFile : error failed to open file %s [%s]", options.Name, err.Error())
+		return handle, err
+	}
+	if !rw.StreamOnly {
+		err = rw.createHandleCache(handle)
+		if err != nil {
+			log.Err("Stream::OpenFile : error failed to create cache object %s [%s]", options.Name, err.Error())
+			return handle, err
+		}
+	}
+	return handle, err
+}
+
+func (rw *ReadWriteCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
+	// log.Trace("Stream::ReadInBuffer : name=%s, handle=%d, offset=%d", options.Handle.Path, options.Handle.ID, options.Offset)
+	options.Handle.CacheObj.Lock()
+	defer options.Handle.CacheObj.Unlock()
+	if !rw.StreamOnly && options.Handle.CacheObj.StreamOnly {
```

Comment:
if the handle is stream only why do we need to create a cache handle?

Reply:
when we convert a handle to stream only it's because we're at capacity, so in subsequent calls we can still check if we're at capacity and if we're not then we can convert to caching on that handle - rather than stream it.